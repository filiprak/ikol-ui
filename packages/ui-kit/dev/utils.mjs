import { existsSync } from 'fs';
import { readFile, writeFile, mkdir } from 'fs/promises';
import path from 'path';
import { fileURLToPath } from 'url';

const __filename = fileURLToPath(import.meta.url);
const __dirname = path.dirname(__filename);

export async function useTemplateFile(filename, outputFilepath, variables) {
  const data = await readFile(path.resolve(__dirname, 'templates', filename), 'utf8');
  return useTemplate(data, outputFilepath, variables);
}

export async function useTemplate(templateStr, outputFilepath, variables) {
  const updatedTemplate = templateStr.replace(/\[\[([a-zA-Z0-9_]+)\]\]/g, (match, key) => {
    return variables[key] || match;
  });

  if (!existsSync(path.dirname(outputFilepath))) {
    await mkdir(path.dirname(outputFilepath));
  }
  await writeFile(outputFilepath, updatedTemplate);
}

export const kebabize = str => {
  return str.split('').map((letter, idx) => {
      return letter.toUpperCase() === letter
          ? `${idx !== 0 ? '-' : ''}${letter.toLowerCase()}`
          : letter;
  }).join('');
}

export const ROOT_DIR = path.resolve(__dirname, '..');
