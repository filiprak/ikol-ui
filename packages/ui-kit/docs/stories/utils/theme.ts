import { useTheme } from '@ui/composables';
import { createIkolUI } from '@ui/index';
import { App } from 'vue';
import { StoryContext, VueRenderer } from '@storybook/vue3';

export function setupTheme(context: any) {
    const theme = useTheme();

    theme.is_dark.value = context?.globals?.theme === 'dark';

    return theme;
}

export function setupApp(app: App, context?: StoryContext<VueRenderer>) {
    app.use(createIkolUI());
}
