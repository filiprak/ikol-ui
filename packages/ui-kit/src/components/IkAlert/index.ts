import IkAlert from './IkAlert.vue';
export interface IkAlert extends InstanceType<typeof IkAlert> { };
export { IkAlert };
