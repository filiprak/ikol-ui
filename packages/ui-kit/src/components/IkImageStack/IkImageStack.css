.ik-image-stack {
    display: inline-flex;
}

.ik-image-stack--view {
    cursor: pointer;
}

.ik-image-stack__image-wrapper {
    background-color: var(--surface-darken-color);
    border: var(--background-color) solid;
    border-radius: 50%;
    position: relative;
    display: flex;
    justify-content: center;
    align-items: center;
}

.ik-image-stack__image-wrapper--disabled .ik-image {
    opacity: .5;
}

.ik-image-stack__image--dark {
    background-color: #000;
    opacity: 0.5;
    border-radius: 50%;
    width: 100%;
    height: 100%;
    position: absolute;
}

.ik-image-stack__count {
    position: absolute;
    color: #fff;
}

.ik-image-stack__add {
    background-color: var(--surface-darken-color);
    box-sizing: content-box;
    color: var(--on-surface-darken-color-3);
}