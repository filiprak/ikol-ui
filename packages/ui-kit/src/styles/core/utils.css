/* shadow */
.ik-shadow-0 {
    box-shadow: none;
}

.ik-shadow-1 {
    box-shadow: 0 1px 3px 0 var(--shadow-color-1);
}

.ik-shadow-2 {
    box-shadow: 0 2px 5px 0 var(--shadow-color-1);
}

.ik-shadow-3 {
    box-shadow: 0 2px 10px 0 var(--shadow-color-1);
}

.ik-shadow-4 {
    box-shadow: 0 2px 15px 0 var(--shadow-color-1);
}

.ik-shadow-5 {
    box-shadow: 0 2px 20px 0 var(--shadow-color-1);
}

.ik-shadow-6 {
    box-shadow: 0 3px 30px 0 var(--shadow-color-1);
}

/* skeleton background */
@-webkit-keyframes skeleton-loading {
    to {
        transform: translateX(100%)
    }
}

@keyframes skeleton-loading {
    to {
        transform: translateX(100%)
    }
}

.ik-bg-skeleton {
    background: var(--skeleton-color-1);
    overflow: hidden;
    position: relative;
    user-select: none;
}

.ik-bg-skeleton:after {
    background: linear-gradient(90deg, hsla(0, 0%, 100%, 0), var(--skeleton-color-2), hsla(0, 0%, 100%, 0));
    -webkit-animation: skeleton-loading 1.5s infinite;
    animation: skeleton-loading 1.5s infinite;
    content: '';
    height: 100%;
    left: 0;
    position: absolute;
    right: 0;
    top: 0;
    transform: translateX(-100%);
    z-index: 1
}

.ik-cursor--pointer {
    cursor: pointer;
}

/* design utils */
.ik-design--bg-primary {
    background-color: var(--primary-color);
    color: var(--on-primary-color);
}

.ik-design--bg-success {
    background-color: var(--success-color);
    color: var(--on-success-color);
}

.ik-design--bg-error {
    background-color: var(--error-color);
    color: var(--on-error-color);
}

.ik-design--bg-accent {
    background-color: var(--accent-color);
    color: var(--on-accent-color);
}

.ik-design--bg-secondary {
    background-color: var(--secondary-color);
    color: var(--on-secondary-color);
}

.ik-design--bg-default {
    background-color: var(--surface-darken-color);
    color: var(--on-surface-darken-color-3);
}

/* height */
.ik-height--full {
    height: 100%;
}

.ik-width--full {
    width: 100%;
}
