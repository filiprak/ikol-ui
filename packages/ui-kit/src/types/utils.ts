import type { ExtractPropTypes } from 'vue';

/* eslint-disable @typescript-eslint/naming-convention */
export type IkExtractProps<T> =
    T extends new (...angs: any) => { $props: infer P; } ? NonNullable<P> : // eslint-disable-line @typescript-eslint/no-explicit-any
    T extends (props: infer P, ...args: any) => any ? P : // eslint-disable-line @typescript-eslint/no-explicit-any
    {};

export type IkExtractPropTypes<T> = Prettify<Readonly<ExtractPropTypes<T>>>;

export type IkExtractExposed<T> =
    T extends new (...angs: any) => infer E ? E : // eslint-disable-line @typescript-eslint/no-explicit-any
    T extends (props: any, ctx: any, expose: (exposed: infer E) => any, ...args: any) => any ? NonNullable<E> : // eslint-disable-line @typescript-eslint/no-explicit-any
    {};

export type IkKeysByType<T, U> = keyof {
    [K in keyof T as T[K] extends U ? K : never]: T[K];
};

export enum IkUIDesignColor {
    primary = 'primary',
    success = 'success',
    error = 'error',
    accent = 'accent',
    default = 'default',
}

export enum IkUIElemSize {
    xs = 'xs',
    sm = 'sm',
    md = 'md',
    lg = 'lg',
}

export enum IkUIVariant {
    flat = 'flat',
    filled = 'filled',
    outline = 'outline',
    pill = 'pill',
}

export enum IkPosition {
    top = 'top',
    right = 'right',
    bottom = 'bottom',
    left = 'left',
    horizontal = 'horizontal',
    vertical = 'vertical',
}

export enum IkAlignment {
    start = 'start',
    center = 'center',
    end = 'end',
}

export type EnumT = Record<string, string> | Record<string | number, string | number>;
export type ObjValueT<T> = T[keyof T];
export type DefinedKeys<T> = { [K in keyof T]: undefined extends T[K] ? never : K }[keyof T];
export type OmitUndefined<T> = { [K in DefinedKeys<T>]: T[K] };
export type NotArray<T> = T extends any[] ? never : T; // eslint-disable-line @typescript-eslint/no-explicit-any
export type DefinedComponent = new (...args: any[]) => any; // eslint-disable-line @typescript-eslint/no-explicit-any
export type Prettify<T> = {
    [K in keyof T]: T[K];
} & {};
