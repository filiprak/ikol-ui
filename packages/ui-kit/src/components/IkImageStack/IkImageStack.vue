<template>
    <div :class="main_class">
        <div v-for="(image, index) in visible_images"
             :key="index"
             :class="imageClass(index)"
             :style="imageStyles(index)">
            <IkImage :src="image"
                     :size="size"
                     round></IkImage>
        </div>
        <div v-if="value?.length > max"
             class="ik-image-stack__image-wrapper"
             :style="imageStyles(max)">
            <IkImage :src="value ? value[max] : undefined"
                     :size="size"
                     round />
            <div class="ik-image-stack__image--dark"></div>
            <span class="ik-image-stack__count ik-noselect"
                  :style="image_count_styles">
                {{ hidden_count }}
            </span>
        </div>
        <div v-else-if="!view_mode"
             class="ik-image-stack__image-wrapper ik-image-stack__add"
             :style="image_placeholder_styles">
            <IkIcon icon="plus" />
        </div>
    </div>
</template>

<script setup lang="ts">
import { IkImage } from '@ui/components/IkImage';
import { IkIcon } from '@ui/components/IkIcon';
import './IkImageStack.css';
import { computed, type CSSProperties, type PropType } from 'vue';

const props = defineProps({
    size: {
        type: Number,
        default: 24,
    },
    max: {
        type: Number,
        default: 5,
    },
    view_mode: {
        type: Boolean,
        default: false,
    },
    value: {
        type: Array as PropType<string[]>,
        default() {
            return [];
        },
    },
    disabled: {
        type: Array as PropType<string[]>,
        default() {
            return [];
        },
    },
});

const visible_images = computed((): string[] => {
    return props.value.slice(0, props.max);
});
const hidden_count = computed((): string => {
    return `+${props.value.length - props.max}`;
});
const image_count_styles = computed((): CSSProperties => {
    return {
        'font-size': `${Math.round(props.size * 0.35)}px`,
    };
});
const image_placeholder_styles = computed((): CSSProperties => {
    return Object.assign(imageStyles(props.value.length), {
        width: `${props.size}px`,
        height: `${props.size}px`,
        'font-size': `${Math.round(props.size * 0.4)}px`,
    });
});
const main_class = computed((): Record<string, boolean> => {
    return {
        'ik-image-stack': true,
        'ik-image-stack--view': !props.view_mode,
    };
});

function getIsDisabled(index: number): boolean {
    return props.disabled.indexOf(props.value[index]) > -1;
}

function imageClass(index: number): Record<string, boolean> {
    return {
        'ik-image-stack__image-wrapper': true,
        'ik-image-stack__image-wrapper--disabled': getIsDisabled(index),
    };
}

function imageStyles(index: number): CSSProperties {
    const margin: number = -Math.round(props.size * 0.3);
    return {
        'margin-left': index !== 0 ? margin + 'px' : '',
        'z-index': props.max - index,
        'border-width': `${Math.round(props.size * 0.06)}px`,
    };
}
</script>
