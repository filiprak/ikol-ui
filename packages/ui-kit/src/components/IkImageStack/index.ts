import IkImageStack from './IkImageStack.vue';
export interface IkImageStack extends InstanceType<typeof IkImageStack> { };
export { IkImageStack };
