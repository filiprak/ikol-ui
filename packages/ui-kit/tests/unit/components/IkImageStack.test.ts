import { renderShallow } from '@ui/tests/tools/utils';
import { IkImageStack } from '@ui/components/IkImageStack';

describe('IkImageStack', () => {
    const getInstance = (options = {}) => renderShallow(IkImageStack, options);

    it('correct default state', () => {
        const wrapper = getInstance();

        expect(wrapper.props()).toMatchObject({ size: 24 });
        expect(wrapper.props()).toMatchObject({ max: 5 });
        expect(wrapper.props()).toMatchObject({ value: [] });
        expect(wrapper.classes()).toStrictEqual([
            'ik-image-stack',
            'ik-image-stack--view',
        ]);
        expect(wrapper.element).toMatchSnapshot();
    });

    it('renders with more then max items', () => {
        const wrapper = getInstance({
            propsData: {
                max: 5,
                value: [
                    'http://test.com/image1',
                    'http://test.com/image2',
                    'http://test.com/image3',
                    'http://test.com/image4',
                    'http://test.com/image5',
                    'http://test.com/image6',
                    'http://test.com/image7',
                ],
            },
        });

        expect(wrapper.vm.visible_images).toStrictEqual([
            'http://test.com/image1',
            'http://test.com/image2',
            'http://test.com/image3',
            'http://test.com/image4',
            'http://test.com/image5',
        ]);

        expect(wrapper.vm.hidden_count).toBe('+2');
        expect(wrapper.findAll('.ik-image-stack__image-wrapper').length).toBe(6);
        expect(wrapper.find('.ik-image-stack__count').attributes().style).toContain('font-size: 8px');
        expect(wrapper.element).toMatchSnapshot();
    });

    it('renders with less then max items', () => {
        const wrapper = getInstance({
            propsData: {
                max: 5,
                value: [
                    'http://test.com/image1',
                    'http://test.com/image2',
                    'http://test.com/image3',
                ],
            },
        });

        expect(wrapper.findAll('.ik-image-stack__image-wrapper').length).toBe(4);
        expect(wrapper.find('.ik-image-stack__add').attributes().style).toContain('width: 24px');
        expect(wrapper.find('.ik-image-stack__add').attributes().style).toContain('height: 24px');
        expect(wrapper.find('.ik-image-stack__add').attributes().style).toContain('font-size: 10px');
        expect(wrapper.element).toMatchSnapshot();
    });

    it('items has correct styles', () => {
        const wrapper = getInstance({
            props: {
                max: 3,
                size: 30,
                value: [
                    'http://test.com/image1',
                    'http://test.com/image2',
                    'http://test.com/image3',
                    'http://test.com/image4',
                ],
            },
        });
        const items = wrapper.findAll('.ik-image-stack__image-wrapper');

        items.forEach((item, index) => {
            expect(item.attributes().style).toContain(index > 1 ? 'margin-left: -9px' : '');
            expect(item.attributes().style).toContain('border-width: 2px');
            expect(item.attributes().style).toContain('z-index: ' + (items.length - index - 1));
        });
    });

    it('renders without add button', () => {
        const wrapper = getInstance({
            propsData: {
                max: 3,
                size: 30,
                view_mode: true,
                value: [
                    'http://test.com/image1',
                    'http://test.com/image2',
                    'http://test.com/image3',
                    'http://test.com/image4',
                ],
            },
        });
        const add_button = wrapper.find('.ik-image-stack__add');

        expect(add_button.exists()).toBeFalsy();

        expect(wrapper.classes()).toStrictEqual([
            'ik-image-stack',
        ]);
    });

    it('renders with disabled images', () => {
        const wrapper = getInstance({
            propsData: {
                max: 5,
                value: [
                    'http://test.com/image1',
                    'http://test.com/image2',
                    'http://test.com/image3',
                ],
                disabled: [
                    'http://test.com/image1',
                ],
            },
        });

        const items = wrapper.findAll('.ik-image-stack__image-wrapper');
        const disabled_item = items[0];

        expect(disabled_item.classes()).toContain('ik-image-stack__image-wrapper--disabled');
    });
});
