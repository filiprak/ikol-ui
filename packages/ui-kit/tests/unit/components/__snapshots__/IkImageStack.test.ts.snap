// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`IkImageStack correct default state 1`] = `
<div
  class="ik-image-stack ik-image-stack--view"
>
  
  
  <div
    class="ik-image-stack__image-wrapper ik-image-stack__add"
    style="z-index: 5; border-width: 1px; width: 24px; height: 24px; font-size: 10px;"
  >
    <ik-icon-stub
      circle="false"
      disabled="false"
      icon="plus"
      tag="i"
    />
  </div>
</div>
`;

exports[`IkImageStack renders with less then max items 1`] = `
<div
  class="ik-image-stack ik-image-stack--view"
>
  
  <div
    class="ik-image-stack__image-wrapper"
    style="z-index: 5; border-width: 1px;"
  >
    <ik-image-stub
      contain="false"
      cover="false"
      loader="false"
      round="true"
      size="24"
      src="http://test.com/image1"
    />
  </div>
  <div
    class="ik-image-stack__image-wrapper"
    style="margin-left: -7px; z-index: 4; border-width: 1px;"
  >
    <ik-image-stub
      contain="false"
      cover="false"
      loader="false"
      round="true"
      size="24"
      src="http://test.com/image2"
    />
  </div>
  <div
    class="ik-image-stack__image-wrapper"
    style="margin-left: -7px; z-index: 3; border-width: 1px;"
  >
    <ik-image-stub
      contain="false"
      cover="false"
      loader="false"
      round="true"
      size="24"
      src="http://test.com/image3"
    />
  </div>
  
  <div
    class="ik-image-stack__image-wrapper ik-image-stack__add"
    style="margin-left: -7px; z-index: 2; border-width: 1px; width: 24px; height: 24px; font-size: 10px;"
  >
    <ik-icon-stub
      circle="false"
      disabled="false"
      icon="plus"
      tag="i"
    />
  </div>
</div>
`;

exports[`IkImageStack renders with more then max items 1`] = `
<div
  class="ik-image-stack ik-image-stack--view"
>
  
  <div
    class="ik-image-stack__image-wrapper"
    style="z-index: 5; border-width: 1px;"
  >
    <ik-image-stub
      contain="false"
      cover="false"
      loader="false"
      round="true"
      size="24"
      src="http://test.com/image1"
    />
  </div>
  <div
    class="ik-image-stack__image-wrapper"
    style="margin-left: -7px; z-index: 4; border-width: 1px;"
  >
    <ik-image-stub
      contain="false"
      cover="false"
      loader="false"
      round="true"
      size="24"
      src="http://test.com/image2"
    />
  </div>
  <div
    class="ik-image-stack__image-wrapper"
    style="margin-left: -7px; z-index: 3; border-width: 1px;"
  >
    <ik-image-stub
      contain="false"
      cover="false"
      loader="false"
      round="true"
      size="24"
      src="http://test.com/image3"
    />
  </div>
  <div
    class="ik-image-stack__image-wrapper"
    style="margin-left: -7px; z-index: 2; border-width: 1px;"
  >
    <ik-image-stub
      contain="false"
      cover="false"
      loader="false"
      round="true"
      size="24"
      src="http://test.com/image4"
    />
  </div>
  <div
    class="ik-image-stack__image-wrapper"
    style="margin-left: -7px; z-index: 1; border-width: 1px;"
  >
    <ik-image-stub
      contain="false"
      cover="false"
      loader="false"
      round="true"
      size="24"
      src="http://test.com/image5"
    />
  </div>
  
  <div
    class="ik-image-stack__image-wrapper"
    style="margin-left: -7px; z-index: 0; border-width: 1px;"
  >
    <ik-image-stub
      contain="false"
      cover="false"
      loader="false"
      round="true"
      size="24"
      src="http://test.com/image6"
    />
    <div
      class="ik-image-stack__image--dark"
    />
    <span
      class="ik-image-stack__count ik-noselect"
      style="font-size: 8px;"
    >
      +2
    </span>
  </div>
</div>
`;
